### ISSUE TITLE
Testing suite for interface contains errors.

### ISSUE THREAD

**Issue by SamMousa** [laravel/framework#11732]
I propose changing the approach for mocking interfaces in the testing suites.
Currently interfaces are mocked in the following way:

        $job = m::mock('Illuminate\Contracts\Queue\Job');

This creates a class that implements the interface, its methods are then mocked using:

        $job->shouldReceive('isDeleted')->once()->andReturn(true);

Mockery will always allow mocking non-existent methods for anonymous classes (a class with no parent class). This means that I can mock any method without it being part of the interface.

When programming against contracts there must not be any other required methods on the concrete classes.

I propose mocking interfaces as follows instead:

````
// Disable mocking of non existent methods.
m::getConfiguration()->allowMockingNonExistentMethods(false);
// Implement interface as subclass of stdClass.
$job = m::mock('stdClass, Illuminate\Contracts\Queue\Job');
````

Doing it this way will make the test itself fail when an attempt is made to mock a method that is not in the interface definition, for example in `QueueWorkerTest::testWorkerLogsJobToFailedQueueIfMaxTriesHasBeenExceeded()`:
````
$job->shouldReceive('getRawBody')->once()->andReturn('body');
````
````
Mockery\Exception: Mockery's configuration currently forbids mocking the method getRawBody as it does not exist on the class or object being mocked
````

Since this will change lots of files (113 occurences of mocking a contract) and will probably break a few tests, I first want to discuss the proposal before creating a PR.

---- End of issue, text below is a note specifically directed at the gatekeeper @GrahamCampbell, not the rest of the team ----
If you do not understand the issue feel free to ask some questions I am more than happy to take the time and dumb it down a bit. Closing the issue because you don't grasp it, like #11720, is not a good approach in the long run.. It doesn't improve the framework and it doesn't actually make you a better programmer. Also it will piss off people and that is often not the best way to get them to make meaningful contributions. Luckily I can be pissed off and think logically at the same time, hence this issue. You are very welcome!

**Comment by GrahamCampbell** [comment]
You already opened an issue, and you suggested a good solution.

**Comment by SamMousa** [comment]
This is a different issue.. it is broader.. can you stop randomly closing stuff if you don't understand it?

**Comment by GrahamCampbell** [comment]
>  $job = m::mock('Illuminate\Contracts\Queue\Job');

Just mock an implementation if you need methods not in the contract.

**Comment by SamMousa** [comment]
You still don't understand... I do not need methods not in the contract, Laravel requires methods not in the contract. Which is bad practice.
And because of the way you write tests you do not find these issues...

**Comment by SamMousa** [comment]
Same thing here #11200 

What are the arguments for allowing extra assumptions outside of the contract?
What value does the contract have if I don't know whether it is complete?

**Comment by EliasZ** [comment]
+1 for reopening. At least have this open for discussion @GrahamCampbell - what is the purpose of a contract if it is non-binding? Might as well leave the contracts out together and typehint concrete implementations all the time...

ping @taylorotwell

**Comment by GrahamCampbell** [comment]
Meh.

**Comment by SamMousa** [comment]
Okay, so currently when an interface is mocked an anonymous class is created.
If instead we create a subclass of `stdClass` instead, tests that have implicit assumptions and are working now because of `shouldReceive()` will all break.

After that we can then change the base class for tests that specifically require it, or alternatively allow mocking non-existent methods for those tests since those are the exceptions.

When that's done we can start removing implicit dependencies where possible, or throwing proper exceptions where not possible.

**Comment by SamMousa** [comment]
This guy..

**Comment by GrahamCampbell** [comment]
Closing since there is no "bug", and nobody else has anything to say here, so as per our policy, this issue has been closed. I almost think we need a bot user to close such issues.
